File: apps/packages/ui/src/components/Option/Watchlists/shared/briefing-contract.ts
import type {
  BriefingPipelineContractV1,
  JobOutputPrefs,
  JobScope,
  WatchlistAudioCast,
  WatchlistFiltersPayload,
  WatchlistJobCreate,
  WatchlistProgramFormat
} from "@/types/watchlists"
import { normalizeWatchlistTemplateName } from "./templateNames"

export type { BriefingPipelineContractV1, WatchlistProgramFormat }

export const BRIEFING_PIPELINE_VERSION = 1 as const
export const DEFAULT_BRIEFING_SELECTION_MAX_ITEMS = 100
export const DEFAULT_BRIEFING_AUDIO_TARGET_MINUTES = 8

export const WATCHLIST_PROGRAM_FORMATS: readonly WatchlistProgramFormat[] = [
  "concise_briefing",
  "solo_update",
  "host_discussion",
  "sportscast",
  "culture_roundtable",
  "custom"
]

const LEGACY_BRIEFING_KEYS = new Set([
  "auto_output",
  "template",
  "template_name",
  "deliveries",
  "delivery_config",
  "generate_audio",
  "target_audio_minutes",
  "audio_language",
  "audio_provider",
  "audio_model",
  "audio_voice",
  "tts_provider",
  "tts_model",
  "tts_voice",
  "audio_speed",
  "audio_cast",
  "voice_map",
  "llm_provider",
  "llm_model",
  "persona_summarize",
  "persona_id",
  "persona_provider",
  "persona_model",
  "background_audio_uri",
  "background_volume",
  "background_delay_ms",
  "background_fade_seconds"
])

type UnknownRecord = Record<string, unknown>

export interface BriefingSetupDraft {
  monitorName: string
  scope: JobScope
  active: boolean
  scheduleExpr?: string | null
  timezone?: string | null
  description?: string
  jobFilters?: WatchlistFiltersPayload
  watchlistId?: number | null
  selectionMode?: "automatic" | "manual_override"
  maxItems?: number
  programFormat?: WatchlistProgramFormat
  outcomeNoun?: "briefing" | "episode"
  showName?: string
  premise?: string
  audience?: string
  tone?: string
  episodeTitlePattern?: string
  customInstructions?: string
  templateName: string
  templateFormat?: "md" | "html"
  templateVersion?: number
  showNotes?: boolean
  audioEnabled: boolean
  targetAudioMinutes?: number
  audioLanguage?: string
  audioProvider?: string
  audioModel?: string
  audioVoice?: string
  audioCast?: WatchlistAudioCast
  voiceMap?: Record<string, string>
  emailEnabled?: boolean
  emailRecipients?: string[]
  chatbookEnabled?: boolean
  chatbookTitle?: string
  preservedOutputPrefs?: JobOutputPrefs | null
}

export interface NormalizedLegacyBriefingContract {
  outputPrefs: JobOutputPrefs
  contract: BriefingPipelineContractV1
  warnings: string[]
}

const isRecord = (value: unknown): value is UnknownRecord =>
  typeof value === "object" && value !== null && !Array.isArray(value)

const cloneValue = <T>(value: T): T => {
  if (Array.isArray(value)) return value.map((entry) => cloneValue(entry)) as T
  if (!isRecord(value)) return value
  return Object.fromEntries(
    Object.entries(value).map(([key, entry]) => [key, cloneValue(entry)])
  ) as T
}

const record = (value: unknown): UnknownRecord =>
  isRecord(value) ? cloneValue(value) : {}

const deepMerge = (base: UnknownRecord, overlay: UnknownRecord): UnknownRecord => {
  const merged = cloneValue(base)
  for (const [key, value] of Object.entries(overlay)) {
    merged[key] = isRecord(value) && isRecord(merged[key])
      ? deepMerge(merged[key] as UnknownRecord, value)
      : cloneValue(value)
  }
  return merged
}

const clampInteger = (value: unknown, fallback: number, min: number, max: number): number => {
  if (
    value === null ||
    value === undefined ||
    (typeof value === "string" && value.trim().length === 0)
  ) {
    return fallback
  }
  const parsed = Number(value)
  const normalized = Number.isFinite(parsed) ? Math.floor(parsed) : fallback
  return Math.min(max, Math.max(min, normalized))
}

const coerceBoolean = (value: unknown, fallback = false): boolean => {
  if (typeof value === "boolean") return value
  const normalized = typeof value === "string" ? value.trim().toLowerCase() : value
  if (normalized === 1 || normalized === "1" || normalized === "true" || normalized === "yes" || normalized === "on") {
    return true
  }
  if (normalized === 0 || normalized === "0" || normalized === "false" || normalized === "no" || normalized === "off" || normalized === "") {
    return false
  }
  return fallback
}

const trimmed = (value: unknown): string =>
  typeof value === "string" ? value.trim() : ""

const normalizeRecipients = (value: unknown): string[] =>
  Array.isArray(value)
    ? Array.from(new Set(value.map((entry) => trimmed(entry)).filter(Boolean)))
    : []

const normalizeVoiceMap = (value: unknown): Record<string, string> | undefined => {
  if (!isRecord(value)) return undefined
  const entries = Object.entries(value)
    .map(([marker, voice]) => [marker.trim(), trimmed(voice)] as const)
    .filter(([marker, voice]) => marker.length > 0 && voice.length > 0)
  return entries.length > 0 ? Object.fromEntries(entries) : undefined
}

const normalizeAudioCast = (value: unknown): WatchlistAudioCast | undefined => {
  if (!isRecord(value) || !Array.isArray(value.speakers)) return undefined
  const speakers = value.speakers
    .filter(isRecord)
    .map((speaker, index) => ({
      id: trimmed(speaker.id) || `speaker_${index + 1}`,
      label: trimmed(speaker.label) || `Speaker ${index + 1}`,
      ...(trimmed(speaker.role) ? { role: trimmed(speaker.role) } : {}),
      voice: trimmed(speaker.voice),
      ...(trimmed(speaker.persona) ? { persona: trimmed(speaker.persona) } : {})
    }))
    .filter((speaker) => speaker.voice.length > 0)
    .slice(0, 4)
  if (speakers.length === 0) return undefined
  return {
    speaker_count: speakers.length as 1 | 2 | 3 | 4,
    speakers
  }
}

const baseContract = (): UnknownRecord => ({
  version: BRIEFING_PIPELINE_VERSION,
  selection: {
    mode: "automatic",
    max_items: DEFAULT_BRIEFING_SELECTION_MAX_ITEMS
  },
  editorial: {
    program_format: "concise_briefing",
    outcome_noun: "briefing"
  },
  text: {
    enabled: true,
    type: "briefing_markdown",
    format: "md",
    template_name: "",
    show_notes: false
  },
  audio: {
    enabled: false,
    language: "en"
  },
  delivery: {
    reports: { enabled: true },
    email: { enabled: false, recipients: [] },
    chatbook: { enabled: false }
  },
  test: {
    external_delivery: false,
    audio_sample_seconds: 60
  }
})

const legacyContract = (source: UnknownRecord): UnknownRecord => {
  const legacy: UnknownRecord = {}
  const autoOutput = record(source.auto_output)
  const template = record(source.template)
  const text: UnknownRecord = { ...autoOutput }
  if (Object.keys(template).length > 0) {
    for (const [key, value] of Object.entries(template)) {
      if (!["default_name", "default_version", "default_format"].includes(key)) {
        text[key] ??= cloneValue(value)
      }
    }
    text.template_name ??= template.default_name
    text.template_version ??= template.default_version
    text.format ??= template.default_format
  }
  if (source.template_name !== undefined) text.template_name = source.template_name
  if (Object.keys(text).length > 0) legacy.text = text

  const audio: UnknownRecord = {}
  if ("generate_audio" in source) audio.enabled = coerceBoolean(source.generate_audio)
  const audioKeys: Record<string, string> = {
    target_audio_minutes: "target_minutes",
    audio_language: "language",
    audio_provider: "provider",
    audio_model: "model",
    audio_voice: "voice",
    audio_speed: "speed",
    audio_cast: "cast",
    voice_map: "voice_map",
    llm_provider: "llm_provider",
    llm_model: "llm_model",
    persona_summarize: "persona_summarize",
    persona_id: "persona_id",
    persona_provider: "persona_provider",
    persona_model: "persona_model",
    background_audio_uri: "background_audio_uri",
    background_volume: "background_volume",
    background_delay_ms: "background_delay_ms",
    background_fade_seconds: "background_fade_seconds"
  }
  for (const [legacyKey, canonicalKey] of Object.entries(audioKeys)) {
    if (source[legacyKey] !== undefined) audio[canonicalKey] = cloneValue(source[legacyKey])
  }
  for (const [canonicalKey, aliases] of Object.entries({
    provider: ["tts_provider"],
    model: ["tts_model"],
    voice: ["tts_voice"]
  })) {
    if (audio[canonicalKey] !== undefined) continue
    const alias = aliases.find((key) => source[key] !== undefined)
    if (alias) audio[canonicalKey] = cloneValue(source[alias])
  }
  if (Object.keys(audio).length > 0) legacy.audio = audio

  const delivery = record(source.deliveries)
  const deliveryConfig = record(source.delivery_config)
  if (Object.keys(deliveryConfig).length > 0) {
    const email = record(delivery.email)
    if (deliveryConfig.email_recipients !== undefined) {
      email.recipients ??= cloneValue(deliveryConfig.email_recipients)
    }
    if (deliveryConfig.email_enabled !== undefined) {
      email.enabled ??= coerceBoolean(deliveryConfig.email_enabled)
    }
    if (Object.keys(email).length > 0) delivery.email = email
    const chatbook = record(delivery.chatbook)
    if (deliveryConfig.create_chatbook !== undefined) {
      chatbook.enabled ??= coerceBoolean(deliveryConfig.create_chatbook)
    }
    if (Object.keys(chatbook).length > 0) delivery.chatbook = chatbook
    for (const [key, value] of Object.entries(deliveryConfig)) {
      if (!["email_recipients", "email_enabled", "create_chatbook"].includes(key)) {
        delivery[key] ??= cloneValue(value)
      }
    }
  }
  if (Object.keys(delivery).length > 0) legacy.delivery = delivery
  return legacy
}

const finalizeContract = (raw: UnknownRecord): BriefingPipelineContractV1 => {
  const selection = record(raw.selection)
  const editorial = record(raw.editorial)
  const text = record(raw.text)
  const audio = record(raw.audio)
  const delivery = record(raw.delivery)
  const email = record(delivery.email)
  const chatbook = record(delivery.chatbook)
  const reports = record(delivery.reports)
  const test = record(raw.test)

  const programFormat = WATCHLIST_PROGRAM_FORMATS.includes(
    editorial.program_format as WatchlistProgramFormat
  )
    ? editorial.program_format as WatchlistProgramFormat
    : "concise_briefing"
  const audioEnabled = coerceBoolean(audio.enabled)
  const cast = normalizeAudioCast(audio.cast)
  const voiceMap = normalizeVoiceMap(audio.voice_map)
  const targetMinutes = audio.target_minutes === undefined && !audioEnabled
    ? undefined
    : clampInteger(
      audio.target_minutes,
      DEFAULT_BRIEFING_AUDIO_TARGET_MINUTES,
      1,
      60
    )
  const finalizedText: UnknownRecord = {
    ...text,
    enabled: true,
    type: "briefing_markdown",
    format: text.format === "html" ? "html" : "md",
    template_name: trimmed(text.template_name),
    show_notes: coerceBoolean(text.show_notes)
  }
  if (Number.isFinite(Number(text.template_version)) && Number(text.template_version) > 0) {
    finalizedText.template_version = Math.floor(Number(text.template_version))
  } else {
    delete finalizedText.template_version
  }
  const finalizedAudio: UnknownRecord = {
    ...audio,
    enabled: audioEnabled,
    language: trimmed(audio.language) || "en"
  }
  if (targetMinutes === undefined) delete finalizedAudio.target_minutes
  else finalizedAudio.target_minutes = targetMinutes
  if (trimmed(audio.voice)) finalizedAudio.voice = trimmed(audio.voice)
  else delete finalizedAudio.voice
  if (cast) finalizedAudio.cast = cast
  else delete finalizedAudio.cast
  if (voiceMap) finalizedAudio.voice_map = voiceMap
  else delete finalizedAudio.voice_map
  const finalizedChatbook: UnknownRecord = {
    ...chatbook,
    enabled: coerceBoolean(chatbook.enabled)
  }
  if (trimmed(chatbook.title)) finalizedChatbook.title = trimmed(chatbook.title)
  else delete finalizedChatbook.title

  return {
    ...raw,
    version: BRIEFING_PIPELINE_VERSION,
    selection: {
      ...selection,
      mode: selection.mode === "manual_override" ? "manual_override" : "automatic",
      max_items: clampInteger(
        selection.max_items,
        DEFAULT_BRIEFING_SELECTION_MAX_ITEMS,
        1,
        1000
      )
    },
    editorial: {
      ...editorial,
      program_format: programFormat,
      outcome_noun: editorial.outcome_noun === "episode" ? "episode" : "briefing"
    },
    text: finalizedText,
    audio: finalizedAudio,
    delivery: {
      ...delivery,
      reports: { ...reports, enabled: true },
      email: {
        ...email,
        enabled: coerceBoolean(email.enabled),
        recipients: normalizeRecipients(email.recipients)
      },
      chatbook: finalizedChatbook
    },
    test: {
      ...test,
      external_delivery: false,
      audio_sample_seconds: 60
    }
  } as BriefingPipelineContractV1
}

export const normalizeLegacyBriefingContract = (
  raw: JobOutputPrefs | null | undefined,
  _options: { scheduled: boolean }
): NormalizedLegacyBriefingContract => {
  const source = record(raw)
  const contract = finalizeContract(
    deepMerge(
      deepMerge(baseContract(), legacyContract(source)),
      record(source.briefing_pipeline)
    )
  )
  const legacyConsumed = Object.keys(source).some((key) => LEGACY_BRIEFING_KEYS.has(key))
  const outputPrefs = cloneValue(source)
  for (const key of LEGACY_BRIEFING_KEYS) delete outputPrefs[key]
  outputPrefs.briefing_pipeline = contract
  return {
    outputPrefs: outputPrefs as JobOutputPrefs,
    contract,
    warnings: legacyConsumed ? ["legacy_briefing_preferences_normalized"] : []
  }
}

const setOptionalString = (
  target: UnknownRecord,
  key: string,
  value: string | undefined
) => {
  if (value === undefined) return
  const normalized = value.trim()
  if (normalized) target[key] = normalized
  else delete target[key]
}

export const buildBriefingPipelineContract = (
  draft: BriefingSetupDraft
): BriefingPipelineContractV1 => {
  const normalized = normalizeLegacyBriefingContract(draft.preservedOutputPrefs, {
    scheduled: Boolean(trimmed(draft.scheduleExpr))
  })
  const existing = normalized.contract
  const preservedContract = record(draft.preservedOutputPrefs?.briefing_pipeline)
  const preservedEditorial = record(preservedContract.editorial)
  const preservedText = record(preservedContract.text)
  const format = draft.programFormat || existing.editorial.program_format
  const preservedOutcomeNoun = preservedEditorial.outcome_noun === "briefing" ||
    preservedEditorial.outcome_noun === "episode"
    ? existing.editorial.outcome_noun
    : undefined
  const outcomeNoun = draft.outcomeNoun || preservedOutcomeNoun || (
      format === "concise_briefing" ? "briefing" : "episode"
    )
  const editorial = record(existing.editorial)
  editorial.program_format = format
  editorial.outcome_noun = outcomeNoun
  setOptionalString(editorial, "show_name", draft.showName)
  setOptionalString(editorial, "premise", draft.premise)
  setOptionalString(editorial, "audience", draft.audience)
  setOptionalString(editorial, "tone", draft.tone)
  setOptionalString(editorial, "episode_title_pattern", draft.episodeTitlePattern)
  setOptionalString(editorial, "custom_instructions", draft.customInstructions)

  const text = record(existing.text)
  const normalizedTemplateName = normalizeWatchlistTemplateName(draft.templateName)
  text.enabled = true
  text.type = "briefing_markdown"
  text.format = draft.templateFormat || existing.text.format || "md"
  text.template_name = normalizedTemplateName
  if (draft.templateVersion !== undefined) {
    if (Number.isFinite(Number(draft.templateVersion)) && Number(draft.templateVersion) > 0) {
      text.template_version = Math.floor(Number(draft.templateVersion))
    } else {
      delete text.template_version
    }
  }
  const preservedShowNotes = Object.prototype.hasOwnProperty.call(
    preservedText,
    "show_notes"
  )
    ? existing.text.show_notes
    : undefined
  text.show_notes = draft.showNotes ?? preservedShowNotes ?? outcomeNoun === "episode"

  const audio = record(existing.audio)
  audio.enabled = Boolean(draft.audioEnabled)
  audio.language = trimmed(draft.audioLanguage) || trimmed(existing.audio.language) || "en"
  setOptionalString(audio, "provider", draft.audioProvider)
  setOptionalString(audio, "model", draft.audioModel)
  if (draft.audioEnabled) {
    audio.target_minutes = clampInteger(
      draft.targetAudioMinutes ?? existing.audio.target_minutes,
      DEFAULT_BRIEFING_AUDIO_TARGET_MINUTES,
      1,
      60
    )
  } else {
    delete audio.target_minutes
  }
  if (draft.audioVoice !== undefined) {
    if (trimmed(draft.audioVoice)) audio.voice = trimmed(draft.audioVoice)
    else delete audio.voice
  }
  if (draft.audioEnabled && !trimmed(audio.voice)) audio.voice = "alloy"
  let cast = draft.audioCast !== undefined
    ? normalizeAudioCast(draft.audioCast)
    : normalizeAudioCast(audio.cast)
  if (draft.audioEnabled && !cast) {
    cast = {
      speaker_count: 1,
      speakers: [
        {
          id: "speaker_1",
          label: "Speaker 1",
          role: "host",
          voice: trimmed(audio.voice) || "alloy"
        }
      ]
    }
  }
  if (cast) audio.cast = cast
  else delete audio.cast
  let voiceMap = draft.voiceMap !== undefined
    ? normalizeVoiceMap(draft.voiceMap)
    : normalizeVoiceMap(audio.voice_map)
  if (draft.audioEnabled && !voiceMap && cast) {
    voiceMap = Object.fromEntries(
      cast.speakers.map((speaker) => [speaker.id, speaker.voice])
    )
  }
  if (voiceMap) audio.voice_map = voiceMap
  else delete audio.voice_map

  const delivery = record(existing.delivery)
  delivery.reports = { ...record(delivery.reports), enabled: true }
  const email = record(delivery.email)
  const emailEnabled = draft.emailEnabled ?? coerceBoolean(email.enabled)
  email.enabled = emailEnabled
  email.recipients = emailEnabled
    ? normalizeRecipients(draft.emailRecipients ?? email.recipients)
    : []
  delivery.email = email
  const chatbook = record(delivery.chatbook)
  const chatbookEnabled = draft.chatbookEnabled ?? coerceBoolean(chatbook.enabled)
  chatbook.enabled = chatbookEnabled
  if (draft.chatbookTitle !== undefined) {
    if (trimmed(draft.chatbookTitle)) chatbook.title = trimmed(draft.chatbookTitle)
    else delete chatbook.title
  }
  delivery.chatbook = chatbook

  return finalizeContract({
    ...record(existing),
    version: BRIEFING_PIPELINE_VERSION,
    selection: {
      ...record(existing.selection),
      mode: draft.selectionMode || existing.selection.mode,
      max_items: clampInteger(
        draft.maxItems ?? existing.selection.max_items,
        DEFAULT_BRIEFING_SELECTION_MAX_ITEMS,
        1,
        1000
      )
    },
    editorial,
    text,
    audio,
    delivery,
    test: {
      ...record(existing.test),
      external_delivery: false,
      audio_sample_seconds: 60
    }
  })
}

export const toCanonicalWatchlistJobPayload = (
  draft: BriefingSetupDraft
): WatchlistJobCreate => {
  const normalized = normalizeLegacyBriefingContract(draft.preservedOutputPrefs, {
    scheduled: Boolean(trimmed(draft.scheduleExpr))
  })
  const contract = buildBriefingPipelineContract(draft)
  const scheduleExpr = trimmed(draft.scheduleExpr) || undefined
  const timezone = trimmed(draft.timezone) || undefined
  const description = trimmed(draft.description) || undefined
  const watchlistId = Number(draft.watchlistId)
  return {
    name: trimmed(draft.monitorName),
    ...(description ? { description } : {}),
    scope: cloneValue(draft.scope),
    active: draft.active,
    schedule_expr: scheduleExpr,
    timezone,
    output_prefs: {
      ...normalized.outputPrefs,
      briefing_pipeline: contract
    },
    ...(draft.jobFilters ? { job_filters: cloneValue(draft.jobFilters) } : {}),
    ...(Number.isFinite(watchlistId) && watchlistId > 0
      ? { watchlist_id: Math.floor(watchlistId) }
      : {})
  }
}

File: tldw_Server_API/app/core/Watchlists/briefing_contract.py
"""Canonical Watchlists briefing preferences and legacy compatibility."""

from __future__ import annotations

import copy
import os
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

BRIEFING_PIPELINE_KEY = "briefing_pipeline"
BRIEFING_PIPELINE_VERSION = 1
PROGRAM_FORMATS = {
    "concise_briefing",
    "solo_update",
    "host_discussion",
    "sportscast",
    "culture_roundtable",
    "custom",
}

_LEGACY_BRIEFING_KEYS = {
    "auto_output",
    "template",
    "template_name",
    "deliveries",
    "delivery_config",
    "generate_audio",
    "target_audio_minutes",
    "audio_language",
    "audio_provider",
    "audio_model",
    "audio_voice",
    "tts_provider",
    "tts_model",
    "tts_voice",
    "audio_speed",
    "audio_cast",
    "voice_map",
    "llm_provider",
    "llm_model",
    "persona_summarize",
    "persona_id",
    "persona_provider",
    "persona_model",
    "background_audio_uri",
    "background_volume",
    "background_delay_ms",
    "background_fade_seconds",
}


@dataclass(frozen=True)
class NormalizedBriefingContract:
    output_prefs: dict[str, Any]
    contract: dict[str, Any]
    warnings: tuple[str, ...]


def _clamped_int(value: Any, default: int, minimum: int, maximum: int) -> int:
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        parsed = default
    return max(minimum, min(maximum, parsed))


def _coerce_bool(value: Any, *, default: bool = False) -> bool:
    """Return a deterministic boolean, falling back safely for invalid values."""
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"true", "1", "yes", "on"}:
            return True
        if normalized in {"false", "0", "no", "off", ""}:
            return False
        return default
    if isinstance(value, (int, float)):
        if value == 1:
            return True
        if value == 0:
            return False
    return default


def _configured_selection_limit() -> int:
    return _clamped_int(os.getenv("WATCHLIST_BRIEFING_MAX_ITEMS"), 100, 1, 1000)


def _mapping_copy(value: Any) -> dict[str, Any]:
    return copy.deepcopy(dict(value)) if isinstance(value, Mapping) else {}


def _deep_merge(base: dict[str, Any], overlay: Mapping[str, Any]) -> dict[str, Any]:
    merged = copy.deepcopy(base)
    for key, value in overlay.items():
        if isinstance(value, Mapping) and isinstance(merged.get(key), dict):
            merged[key] = _deep_merge(merged[key], value)
        else:
            merged[key] = copy.deepcopy(value)
    return merged


def _base_contract(*, scheduled: bool) -> dict[str, Any]:
    return {
        "version": BRIEFING_PIPELINE_VERSION,
        "selection": {"mode": "automatic", "max_items": _configured_selection_limit()},
        "editorial": {"program_format": "concise_briefing", "outcome_noun": "briefing"},
        "text": {
            "enabled": _coerce_bool(scheduled),
            "type": "briefing_markdown",
            "format": "md",
            "template_name": "",
            "show_notes": False,
        },
        "audio": {
            "enabled": False,
            "target_minutes": 10,
            "language": "en",
        },
        "delivery": {
            "reports": {"enabled": True},
            "email": {"enabled": False, "recipients": []},
            "chatbook": {"enabled": False},
        },
        "test": {"external_delivery": False, "audio_sample_seconds": 60},
    }


def _legacy_contract(raw: Mapping[str, Any]) -> dict[str, Any]:
    legacy: dict[str, Any] = {}

    auto_output = _mapping_copy(raw.get("auto_output"))
    template = _mapping_copy(raw.get("template"))
    text = auto_output
    if template:
        for key, value in template.items():
            if key not in {"default_name", "default_version", "default_format"}:
                text.setdefault(key, value)
        text.setdefault("template_name", template.get("default_name"))
        text.setdefault("template_version", template.get("default_version"))
        text.setdefault("format", template.get("default_format"))
    if raw.get("template_name") is not None:
        text["template_name"] = copy.deepcopy(raw.get("template_name"))
    text = {key: value for key, value in text.items() if value is not None}
    if text:
        legacy["text"] = text

    audio_key_map = {
        "target_audio_minutes": "target_minutes",
        "audio_language": "language",
        "audio_provider": "provider",
        "audio_model": "model",
        "audio_voice": "voice",
        "audio_speed": "speed",
        "audio_cast": "cast",
        "voice_map": "voice_map",
        "llm_provider": "llm_provider",
        "llm_model": "llm_model",
        "persona_summarize": "persona_summarize",
        "persona_id": "persona_id",
        "persona_provider": "persona_provider",
        "persona_model": "persona_model",
        "background_audio_uri": "background_audio_uri",
        "background_volume": "background_volume",
        "background_delay_ms": "background_delay_ms",
        "background_fade_seconds": "background_fade_seconds",
    }
    audio: dict[str, Any] = {}
    if "generate_audio" in raw:
        audio["enabled"] = _coerce_bool(raw.get("generate_audio"))
    for old_key, new_key in audio_key_map.items():
        if old_key in raw and raw.get(old_key) is not None:
            audio[new_key] = copy.deepcopy(raw.get(old_key))
    for canonical_key, aliases in {
        "provider": ("tts_provider",),
        "model": ("tts_model",),
        "voice": ("tts_voice",),
    }.items():
        if canonical_key not in audio:
            for alias in aliases:
                if raw.get(alias) is not None:
                    audio[canonical_key] = copy.deepcopy(raw.get(alias))
                    break
    if audio:
        legacy["audio"] = audio

    delivery = _mapping_copy(raw.get("deliveries"))
    delivery_config = _mapping_copy(raw.get("delivery_config"))
    if delivery_config:
        email = _mapping_copy(delivery.get("email"))
        if "email_recipients" in delivery_config:
            email.setdefault("recipients", copy.deepcopy(delivery_config["email_recipients"]))
        if "email_enabled" in delivery_config:
            email.setdefault("enabled", _coerce_bool(delivery_config["email_enabled"]))
        if email:
            delivery["email"] = email
        chatbook = _mapping_copy(delivery.get("chatbook"))
        if "create_chatbook" in delivery_config:
            chatbook.setdefault("enabled", _coerce_bool(delivery_config["create_chatbook"]))
        if chatbook:
            delivery["chatbook"] = chatbook
        for key, value in delivery_config.items():
            if key not in {"email_recipients", "email_enabled", "create_chatbook"}:
                delivery.setdefault(key, copy.deepcopy(value))
    if delivery:
        legacy["delivery"] = delivery

    return legacy


def _finalize_contract(contract: dict[str, Any], *, scheduled: bool) -> dict[str, Any]:
    defaults = _base_contract(scheduled=scheduled)
    for section in ("selection", "editorial", "text", "audio", "delivery", "test"):
        value = contract.get(section)
        contract[section] = (
            _deep_merge(defaults[section], value) if isinstance(value, Mapping) else copy.deepcopy(defaults[section])
        )

    contract["version"] = BRIEFING_PIPELINE_VERSION
    selection = contract["selection"]
    selection["max_items"] = _clamped_int(selection.get("max_items"), _configured_selection_limit(), 1, 1000)
    if selection.get("mode") not in {"automatic", "manual_override"}:
        selection["mode"] = "automatic"

    editorial = contract["editorial"]
    if editorial.get("program_format") not in PROGRAM_FORMATS:
        editorial["program_format"] = "concise_briefing"
    if editorial.get("outcome_noun") not in {"briefing", "episode"}:
        editorial["outcome_noun"] = "briefing"

    audio = contract["audio"]
    audio["enabled"] = _coerce_bool(audio.get("enabled"))
    if "persona_summarize" in audio:
        audio["persona_summarize"] = _coerce_bool(audio.get("persona_summarize"))
    audio["target_minutes"] = _clamped_int(audio.get("target_minutes"), 10, 1, 60)
    audio["language"] = str(audio.get("language") or "en")

    text = contract["text"]
    text["enabled"] = _coerce_bool(scheduled) or _coerce_bool(text.get("enabled")) or audio["enabled"]
    text["type"] = str(text.get("type") or "briefing_markdown")
    text["format"] = text.get("format") if text.get("format") in {"md", "html"} else "md"
    text["template_name"] = str(text.get("template_name") or "")
    text["show_notes"] = _coerce_bool(text.get("show_notes"))

    delivery = contract["delivery"]
    delivery["reports"] = _deep_merge(_mapping_copy(delivery.get("reports")), {"enabled": True})
    for channel in ("email", "chatbook"):
        channel_config = _mapping_copy(delivery.get(channel))
        channel_config["enabled"] = _coerce_bool(channel_config.get("enabled"))
        delivery[channel] = channel_config
    if not isinstance(delivery["email"].get("recipients"), list):
        delivery["email"]["recipients"] = []

    test = contract["test"]
    test["external_delivery"] = False
    test["audio_sample_seconds"] = 60
    return contract


def normalize_briefing_output_prefs(
    raw: Mapping[str, Any] | None,
    *,
    scheduled: bool,
) -> NormalizedBriefingContract:
    """Return canonical briefing intent while preserving unrelated fields."""
    source = _mapping_copy(raw)
    canonical = _mapping_copy(source.get(BRIEFING_PIPELINE_KEY))
    contract = _base_contract(scheduled=scheduled)
    contract = _deep_merge(contract, _legacy_contract(source))
    contract = _deep_merge(contract, canonical)
    contract = _finalize_contract(contract, scheduled=scheduled)

    legacy_consumed = any(key in source for key in _LEGACY_BRIEFING_KEYS)
    output_prefs = copy.deepcopy(source)
    for key in _LEGACY_BRIEFING_KEYS:
        output_prefs.pop(key, None)
    output_prefs[BRIEFING_PIPELINE_KEY] = copy.deepcopy(contract)
    warnings = ("legacy_briefing_preferences_normalized",) if legacy_consumed else ()
    return NormalizedBriefingContract(output_prefs=output_prefs, contract=contract, warnings=warnings)


def get_briefing_contract(
    raw: Mapping[str, Any] | None,
    *,
    scheduled: bool,
) -> dict[str, Any]:
    """Read canonical or legacy briefing preferences without mutating input."""
    return normalize_briefing_output_prefs(raw, scheduled=scheduled).contract


def briefing_selection_limit(contract: Mapping[str, Any]) -> int:
    """Return max_items clamped to the configured 1..1000 range."""
    selection = contract.get("selection")
    value = selection.get("max_items") if isinstance(selection, Mapping) else None
    return _clamped_int(value, _configured_selection_limit(), 1, 1000)
